src/super_gradients/training/datasets/detection_datasets/detection_dataset_v2.py

```diff
@@ -0,0 +1,336 @@
+
+    This is a boilerplate class to facilitate the implementation of datasets.
+
+    HOW TO CREATE A DATASET THAT INHERITS FROM DetectionDataSetV2 ?
+        - Inherit from DetectionDataSetV2
+        - implement the method self._load_annotation to return at least the fields "target" and "img_path"
+        - Call super().__init__ with the required params.
+                //!\\ super().__init__ will call self._load_annotation, so make sure that every required
+                      attributes are set up before calling super().__init__ (ideally just call it last)
+
+    WORKFLOW:
+        - On instantiation:
+            - All annotations are cached. If class_inclusion_list was specified, there is also subclassing at this step.
+            - If cache is True, the images are also cached
+
+        - On call (__getitem__) for a specific image index:
+            - The image and annotations are grouped together in a dict called SAMPLE
+            - the sample is processed according to th transform
+            - Only the specified fields are returned by __getitem__
+
+    TERMINOLOGY
+        - TARGET:       Groundtruth, made of bboxes. The format can vary from one dataset to another
+        - ANNOTATION:   Combination of targets (groundtruth) and metadata of the image, but without the image itself.
+                            > Has to include the fields "target" and "img_path"
+                            > Can include other fields like "crowd_target", "image_info", "segmentation", ...
+        - SAMPLE:       Outout of the dataset:
+                            > Has to include the fields "target" and "image"
+                            > Can include other fields like "crowd_target", "image_info", "segmentation", ...
+        - INDEX:        Refers to the index in the dataset.
+        - SAMPLE ID:    Refers to the id of sample before droping any annotaion.
+                            Let's imagine a situation where the downloaded data is made of 120 images but 20 were drop
+                            because they had no annotation. In that case:
+                                > We have 120 samples so sample_id will be between 0 and 119
+                                > But only 100 will be indexed so index will be between 0 and 99
+                                > Therefore, we also have len(self) = 100
+    """
+
+    def __init__(
+            self,
+            input_dim: tuple,
+            n_available_samples: int,
+            target_format: DetectionTargetsFormat,
+            cache: bool = False,
+            cache_path: str = None,
+            transforms: List[DetectionTransform] = [],
+            all_classes_list: Optional[List[str]] = None,
+            class_inclusion_list: Optional[List[str]] = None,
+            ignore_empty_annotations: bool = True,
+            annotation_fields_to_subclass: List[str] = None,
+            output_fields: List[str] = None,
+            collate_fn: Callable = None,
```

Comment:
I was not sure about the clean way to do, but from what I've seen this is how every dataset work except cocov2:
- The dataset has a collate_fn attribute
- The dataset interface uses it in build_data_loaders

Check line 163 of dataset_interface.py: https://github.com/Deci-AI/super-gradients/blob/3698e4b79689edf861ba5e893ab1296ac2674aa2/src/super_gradients/training/datasets/dataset_interfaces/dataset_interface.py
`
        # WRAPPING collate_fn
        train_collate_fn = core_utils.get_param(self.trainset, 'collate_fn')
        val_collate_fn = core_utils.get_param(self.valset, 'collate_fn')
        test_collate_fn = core_utils.get_param(self.testset, 'collate_fn')
`

Reply:
So it is used, but maybe we want to not use this approach and to instead have something similar to CocoDatasetInterfaceV2 ?
I personally don't like the current approach that I coded because as you saw it's hard to know where collate_fn is used which complexifies the code.

Reply:
V2 approach is better.

Reply:
Ok, I'll do it

Reply:
https://github.com/Deci-AI/super-gradients/blob/484cb55e685cbcf946b23da424de198834d48f2c/src/super_gradients/training/datasets/dataset_interfaces/dataset_interface.py#L914-L945